File: tests/wasm/test_bollinger_bands_width.js
/**
 * WASM binding tests for Bollinger Bands Width indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('BBW partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.bollinger_bands_width_js(
        close,
        22,     
        2.2,    
        2.0,    
        "ema",  
        null    
    );
    
    assert.strictEqual(result.length, close.length);
});

test('BBW default params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.bollinger_bands_width_js(
        close,
        20,     
        2.0,    
        2.0,    
        null,   
        null    
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    assertAllNaN(result.slice(0, 19), "Expected NaN in warmup period");
    
    
    if (result.length > 240) {
        assertNoNaN(result.slice(240), "Expected no NaN after sufficient warmup");
    }
});

test('BBW accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.bollinger_bands_width || {};
    
    const result = wasm.bollinger_bands_width_js(
        close,
        expected.default_params?.period || 20,
        expected.default_params?.devup || 2.0,
        expected.default_params?.devdn || 2.0,
        expected.default_params?.matype || "sma",
        expected.default_params?.devtype || 0
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    if (expected.last_5_values) {
        const last5 = result.slice(-5);
        assertArrayClose(
            last5,
            expected.last_5_values,
            1e-8,
            "BBW last 5 values mismatch"
        );
    }
    
    
    await compareWithRust('bollinger_bands_width', result, 'close', {
        period: expected.default_params?.period || 20,
        devup: expected.default_params?.devup || 2.0,
        devdn: expected.default_params?.devdn || 2.0,
        matype: expected.default_params?.matype || "sma",
        devtype: expected.default_params?.devtype || 0
    });
});

test('BBW zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.bollinger_bands_width_js(inputData, 0, 2.0, 2.0, null, null);
    }, /Invalid period|period/);
});

test('BBW period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.bollinger_bands_width_js(dataSmall, 10, 2.0, 2.0, null, null);
    }, /Invalid period|period/);
});

test('BBW very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.bollinger_bands_width_js(singlePoint, 20, 2.0, 2.0, null, null);
    }, /Invalid period|Not enough valid data/);
});

test('BBW empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.bollinger_bands_width_js(empty, 20, 2.0, 2.0, null, null);
    }, /Empty data/);
});

test('BBW all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.bollinger_bands_width_js(allNaN, 20, 2.0, 2.0, null, null);
    }, /All values are NaN/);
});

test('BBW different MA types', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    const matypes = ["sma", "ema", "wma", "dema", "tema"];
    
    for (const matype of matypes) {
        const result = wasm.bollinger_bands_width_js(
            close,
            14,
            2.0,
            2.0,
            matype,
            0
        );
        
        assert.strictEqual(result.length, 100);
        
        for (let i = 0; i < 13; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} for ${matype}`);
        }
    }
});

test('BBW different deviation types', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    const devtypes = [0, 1, 2]; 
    
    for (const devtype of devtypes) {
        const result = wasm.bollinger_bands_width_js(
            close,
            14,
            2.0,
            2.0,
            "sma",
            devtype
        );
        
        assert.strictEqual(result.length, 100);
    }
});

test('BBW NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.bollinger_bands_width_js(close, 20, 2.0, 2.0, null, null);
    assert.strictEqual(result.length, close.length);
    
    
    assertAllNaN(result.slice(0, 19), "Expected NaN in warmup period");
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('BBW batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.bollinger_bands_width_batch_js(
        close,
        20, 20, 0,      
        2.0, 2.0, 0,    
        2.0, 2.0, 0     
    );
    
    
    const singleResult = wasm.bollinger_bands_width_js(close, 20, 2.0, 2.0, null, null);
    
    assert.strictEqual(batchResult.length, singleResult.length);
    assertArrayClose(batchResult, singleResult, 1e-10, "Batch vs single mismatch");
});

test('BBW batch multiple parameters', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.bollinger_bands_width_batch_js(
        close,
        10, 30, 10,     
        1.5, 2.5, 0.5,  
        2.0, 2.0, 0     
    );
    
    
    assert.strictEqual(batchResult.length, 9 * 100);
    
    
    const periods = [10, 20, 30];
    const devups = [1.5, 2.0, 2.5];
    let rowIdx = 0;
    
    for (const period of periods) {
        for (const devup of devups) {
            const rowStart = rowIdx * 100;
            const rowEnd = rowStart + 100;
            const rowData = batchResult.slice(rowStart, rowEnd);
            
            const singleResult = wasm.bollinger_bands_width_js(
                close, period, devup, 2.0, null, null
            );
            
            assertArrayClose(
                rowData, 
                singleResult, 
                1e-10, 
                `Period ${period}, devup ${devup} mismatch`
            );
            
            rowIdx++;
        }
    }
});

test('BBW batch metadata', () => {
    
    const metadata = wasm.bollinger_bands_width_batch_metadata_js(
        10, 20, 10,     
        1.5, 2.0, 0.5,  
        2.0, 3.0, 1.0   
    );
    
    
    
    assert.strictEqual(metadata.length, 8 * 3);
    
    
    assert.strictEqual(metadata[0], 10);   
    assert.strictEqual(metadata[1], 1.5);  
    assert.strictEqual(metadata[2], 2.0);  
    
    
    assert.strictEqual(metadata[21], 20);  
    assert.strictEqual(metadata[22], 2.0); 
    assert.strictEqual(metadata[23], 3.0); 
});

test('BBW batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.bollinger_bands_width_batch_js(
        close,
        10, 15, 5,      
        2.0, 2.5, 0.5,  
        1.5, 1.5, 0     
    );
    
    const metadata = wasm.bollinger_bands_width_batch_metadata_js(
        10, 15, 5,
        2.0, 2.5, 0.5,
        1.5, 1.5, 0
    );
    
    
    const numCombos = metadata.length / 3;
    assert.strictEqual(numCombos, 4);
    assert.strictEqual(batchResult.length, 4 * 50);
    
    
    for (let combo = 0; combo < numCombos; combo++) {
        const period = metadata[combo * 3];
        const devup = metadata[combo * 3 + 1];
        const devdn = metadata[combo * 3 + 2];
        
        const rowStart = combo * 50;
        const rowData = batchResult.slice(rowStart, rowStart + 50);
        
        
        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }
        
        
        for (let i = period - 1; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('BBW batch edge cases', () => {
    
    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    
    
    const singleBatch = wasm.bollinger_bands_width_batch_js(
        close,
        5, 5, 1,
        2.0, 2.0, 0.1,
        2.0, 2.0, 1.0
    );
    
    assert.strictEqual(singleBatch.length, 10);
    
    
    const largeBatch = wasm.bollinger_bands_width_batch_js(
        close,
        5, 7, 10, 
        2.0, 2.0, 0,
        2.0, 2.0, 0
    );
    
    
    assert.strictEqual(largeBatch.length, 10);
    
    
    assert.throws(() => {
        wasm.bollinger_bands_width_batch_js(
            new Float64Array([]),
            10, 10, 0,
            2.0, 2.0, 0,
            2.0, 2.0, 0
        );
    }, /All values are NaN/);
});


test('BBW batch - new ergonomic API with single parameter', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.bollinger_bands_width_batch(close, {
        period_range: [20, 20, 0],
        devup_range: [2.0, 2.0, 0],
        devdn_range: [2.0, 2.0, 0],
        matype: "sma",
        devtype: 0
    });
    
    
    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');
    
    
    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, close.length);
    
    
    const combo = result.combos[0];
    assert.strictEqual(combo.period, 20);
    assert.strictEqual(combo.devup, 2.0);
    assert.strictEqual(combo.devdn, 2.0);
    assert.strictEqual(combo.matype, "sma");
    assert.strictEqual(combo.devtype, 0);
    
    
    const oldResult = wasm.bollinger_bands_width_js(close, 20, 2.0, 2.0, "sma", 0);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(result.values[i])) {
            continue; 
        }
        assert(Math.abs(oldResult[i] - result.values[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('BBW batch - new API with multiple parameters', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const result = wasm.bollinger_bands_width_batch(close, {
        period_range: [10, 20, 10],     
        devup_range: [1.5, 2.0, 0.5],   
        devdn_range: [2.0, 2.0, 0],     
        matype: "ema",
        devtype: 1
    });
    
    
    assert.strictEqual(result.rows, 4);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 4);
    assert.strictEqual(result.values.length, 200);
    
    
    const expectedCombos = [
        { period: 10, devup: 1.5, devdn: 2.0 },
        { period: 10, devup: 2.0, devdn: 2.0 },
        { period: 20, devup: 1.5, devdn: 2.0 },
        { period: 20, devup: 2.0, devdn: 2.0 }
    ];
    
    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.combos[i].period, expectedCombos[i].period);
        assert.strictEqual(result.combos[i].devup, expectedCombos[i].devup);
        assert.strictEqual(result.combos[i].devdn, expectedCombos[i].devdn);
        assert.strictEqual(result.combos[i].matype, "ema");
        assert.strictEqual(result.combos[i].devtype, 1);
    }
});

test('BBW batch - new API error handling', () => {
    const close = new Float64Array(testData.close.slice(0, 10));
    
    
    assert.throws(() => {
        wasm.bollinger_bands_width_batch(close, {
            period_range: [10, 10], 
            devup_range: [2.0, 2.0, 0],
            devdn_range: [2.0, 2.0, 0]
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.bollinger_bands_width_batch(close, {
            period_range: [10, 10, 0],
            devup_range: [2.0, 2.0, 0]
            
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.bollinger_bands_width_batch(close, {
            period_range: "invalid",
            devup_range: [2.0, 2.0, 0],
            devdn_range: [2.0, 2.0, 0]
        });
    }, /Invalid config/);
});

test.after(() => {
    console.log('Bollinger Bands Width WASM tests completed');
});

File: tests/python/test_bollinger_bands_width.py
"""
Python binding tests for Bollinger Bands Width indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import os
import sys

# Add the parent directory to the Python path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))
import my_project

from test_utils import (
    load_test_data,
    assert_close,
    assert_array_close,
    assert_all_nan,
    assert_no_nan,
    EXPECTED_OUTPUTS
)


class TestBollingerBandsWidth:
    """Test suite for Bollinger Bands Width Python bindings."""
    
    @classmethod
    def setup_class(cls):
        """Load test data once for all tests."""
        cls.data = load_test_data()
        cls.close = np.array(cls.data['close'], dtype=np.float64)
        cls.high = np.array(cls.data['high'], dtype=np.float64)
        cls.low = np.array(cls.data['low'], dtype=np.float64)
        cls.expected = EXPECTED_OUTPUTS.get('bollinger_bands_width', {})
    
    def test_partial_params(self):
        """Test BBW with partial parameters - mirrors check_bbw_partial_params."""
        # Custom period and devup, default others
        result = my_project.bollinger_bands_width(
            self.close,
            period=22,
            devup=2.2,
            devdn=2.0,  # default
            matype="ema",
            devtype=None  # use default
        )
        
        assert len(result) == len(self.close)
        assert isinstance(result, np.ndarray)
        assert result.dtype == np.float64
    
    def test_default_params(self):
        """Test BBW with default parameters - mirrors check_bbw_default."""
        # Use default params
        result = my_project.bollinger_bands_width(
            self.close,
            period=20,
            devup=2.0,
            devdn=2.0
        )
        
        assert len(result) == len(self.close)
        
        # Check warmup period
        assert_all_nan(result[:19], "Expected NaN in warmup period")
        
        # After warmup, should have values
        assert_no_nan(result[240:], "Expected no NaN after sufficient warmup")
    
    def test_accuracy(self):
        """Test BBW accuracy with expected values."""
        result = my_project.bollinger_bands_width(
            self.close,
            period=self.expected.get('default_params', {}).get('period', 20),
            devup=self.expected.get('default_params', {}).get('devup', 2.0),
            devdn=self.expected.get('default_params', {}).get('devdn', 2.0)
        )
        
        assert len(result) == len(self.close)
        
        # If we have expected values, check them
        if 'last_5_values' in self.expected:
            last_5 = result[-5:]
            assert_array_close(
                last_5,
                self.expected['last_5_values'],
                rtol=1e-8,
                msg="BBW last 5 values mismatch"
            )
    
    def test_zero_period(self):
        """Test BBW fails with zero period - mirrors check_bbw_zero_period."""
        with pytest.raises(ValueError, match="period"):
            my_project.bollinger_bands_width(
                np.array([10.0, 20.0, 30.0]),
                period=0,
                devup=2.0,
                devdn=2.0
            )
    
    def test_period_exceeds_length(self):
        """Test BBW fails when period exceeds data length - mirrors check_bbw_period_exceeds_length."""
        small_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="period"):
            my_project.bollinger_bands_width(
                small_data,
                period=10,
                devup=2.0,
                devdn=2.0
            )
    
    def test_very_small_dataset(self):
        """Test BBW fails with insufficient data - mirrors check_bbw_very_small_dataset."""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError):
            my_project.bollinger_bands_width(
                single_point,
                period=20,
                devup=2.0,
                devdn=2.0
            )
    
    def test_empty_input(self):
        """Test BBW fails with empty input."""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty"):
            my_project.bollinger_bands_width(
                empty,
                period=20,
                devup=2.0,
                devdn=2.0
            )
    
    def test_all_nan_input(self):
        """Test BBW with all NaN values."""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            my_project.bollinger_bands_width(
                all_nan,
                period=20,
                devup=2.0,
                devdn=2.0
            )
    
    def test_different_matypes(self):
        """Test BBW with different moving average types."""
        matypes = ["sma", "ema", "wma", "dema", "tema"]
        
        for matype in matypes:
            result = my_project.bollinger_bands_width(
                self.close[:100],  # Use smaller dataset
                period=14,
                devup=2.0,
                devdn=2.0,
                matype=matype
            )
            
            assert len(result) == 100
            assert isinstance(result, np.ndarray)
    
    def test_different_devtypes(self):
        """Test BBW with different deviation types."""
        devtypes = [0, 1, 2]  # stddev, mean_ad, median_ad
        
        for devtype in devtypes:
            result = my_project.bollinger_bands_width(
                self.close[:100],  # Use smaller dataset
                period=14,
                devup=2.0,
                devdn=2.0,
                devtype=devtype
            )
            
            assert len(result) == 100
            assert isinstance(result, np.ndarray)
    
    def test_nan_handling(self):
        """Test BBW handles NaN values correctly."""
        result = my_project.bollinger_bands_width(
            self.close,
            period=20,
            devup=2.0,
            devdn=2.0
        )
        
        assert len(result) == len(self.close)
        
        # First period-1 values should be NaN
        assert_all_nan(result[:19], "Expected NaN in warmup period")
        
        # After a reasonable warmup, no NaN values should exist
        if len(result) > 240:
            assert_no_nan(result[240:], "Found unexpected NaN after warmup")
    
    def test_streaming_api(self):
        """Test streaming BBW calculation."""
        period = 20
        stream = my_project.BollingerBandsWidthStream(
            period=period,
            devup=2.0,
            devdn=2.0,
            matype="sma",  # Default moving average type
            devtype=0      # Default deviation type (stddev)
        )
        
        # Feed data one by one
        results = []
        for i, val in enumerate(self.close[:50]):
            result = stream.update(val)
            results.append(result)
        
        # First period-1 updates should return None
        for i in range(period - 1):
            assert results[i] is None, f"Expected None at index {i}"
        
        # After that, should get values
        for i in range(period - 1, 50):
            assert results[i] is not None, f"Expected value at index {i}"
            assert not np.isnan(results[i]), f"Got NaN at index {i}"
    
    def test_batch_single_params(self):
        """Test batch processing with single parameter set."""
        # Single parameter combination
        result = my_project.bollinger_bands_width_batch(
            self.close,
            period_range=(20, 20, 0),
            devup_range=(2.0, 2.0, 0),
            devdn_range=(2.0, 2.0, 0)
        )
        
        assert 'values' in result
        assert 'periods' in result
        assert 'devups' in result
        assert 'devdns' in result
        
        # Should have 1 row
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(self.close)
        
        # Compare with single calculation
        single_result = my_project.bollinger_bands_width(
            self.close,
            period=20,
            devup=2.0,
            devdn=2.0
        )
        
        assert_array_close(
            result['values'][0],
            single_result,
            rtol=1e-10,
            msg="Batch vs single mismatch"
        )
    
    def test_batch_multiple_params(self):
        """Test batch processing with multiple parameter combinations."""
        # Multiple periods and deviations
        result = my_project.bollinger_bands_width_batch(
            self.close[:100],  # Use smaller dataset
            period_range=(10, 30, 10),  # 10, 20, 30
            devup_range=(1.5, 2.5, 0.5),  # 1.5, 2.0, 2.5
            devdn_range=(2.0, 2.0, 0)  # 2.0
        )
        
        # Should have 3 * 3 * 1 = 9 combinations
        assert result['values'].shape[0] == 9
        assert result['values'].shape[1] == 100
        
        # Check metadata arrays
        assert len(result['periods']) == 9
        assert len(result['devups']) == 9
        assert len(result['devdns']) == 9
        
        # Verify first combination
        assert result['periods'][0] == 10
        assert result['devups'][0] == 1.5
        assert result['devdns'][0] == 2.0
    
    def test_kernel_selection(self):
        """Test different kernel selections."""
        kernels = [None, "scalar", "avx2", "avx512"]
        
        for kernel in kernels:
            try:
                result = my_project.bollinger_bands_width(
                    self.close[:100],
                    period=20,
                    devup=2.0,
                    devdn=2.0,
                    kernel=kernel
                )
                assert len(result) == 100
            except ValueError as e:
                # Some kernels might not be supported on this platform
                assert "not supported" in str(e) or "Unsupported" in str(e)
    
    def test_edge_cases(self):
        """Test edge cases for BBW."""
        # Very small period
        result = my_project.bollinger_bands_width(
            self.close[:10],
            period=2,
            devup=2.0,
            devdn=2.0
        )
        assert len(result) == 10
        assert not np.isnan(result[1])  # Should have value after period-1
        
        # Large deviations
        result = my_project.bollinger_bands_width(
            self.close[:50],
            period=10,
            devup=5.0,
            devdn=5.0
        )
        assert len(result) == 50
        
        # Asymmetric deviations
        result = my_project.bollinger_bands_width(
            self.close[:50],
            period=10,
            devup=3.0,
            devdn=1.0
        )
        assert len(result) == 50


if __name__ == "__main__":
    pytest.main([__file__])